lib/rpmsg/rpmsg_virtio.c

```diff
@@ -23,6 +23,35 @@
 /* Time to wait - In multiple of 10 msecs. */
 #define RPMSG_TICKS_PER_INTERVAL                10
 
+#define WORD_SIZE                sizeof(unsigned long)
+#define WORD_ALIGN(a)            (((a) & (WORD_SIZE-1)) != 0)? \
+                                 (((a) & (~(WORD_SIZE-1))) + WORD_SIZE):(a)
+
+static void *
+rpmsg_virtio_shm_pool_get_buffer(struct rpmsg_virtio_shm_pool *shpool,
+				 size_t size)
+{
+	void *buffer;
+
+	if(shpool->avail < size)
+		return NULL;
+	buffer =  shpool->base + shpool->size - shpool->avail;
+	shpool->avail -= size;
+
+	return buffer;
+}
+
+void rpmsg_virtio_init_shm_pool(struct rpmsg_virtio_shm_pool *shpool,
+				void *shb, size_t size)
+{
+	if (!shpool)
+		return;
+	shpool->base = shb;
+	shpool->size = WORD_ALIGN(size);
+	shpool->avail = WORD_ALIGN(size);
+	shpool->get_buffer = rpmsg_virtio_shm_pool_get_buffer;
+}
```

Comment:
what about  a simple macro?
application declare it with the macro:
struct rpmsg_virtio_shm_pool shpool = RPMSG_VIRTIO_INIT_SHM_POOL( *shmem, size, NULL);

else you need to return a status on error.